File: src/python/spherical_norm.py
import numpy as np
import math
import cv2
from scipy.spatial.transform import Rotation as R
import numpy as np

def bbox_to_spherical_coord(bbox_center_x, bbox_center_y, frame_width, frame_height):
    u = bbox_center_x / frame_width
    v = bbox_center_y / frame_height
    
    longitude = (u * 2 * math.pi) - math.pi  # [-π, π]
    latitude = (math.pi / 2) - (v * math.pi)  # [π/2, -π/2]
    
    return longitude, latitude


def spherical_to_cartesian(longitude, latitude, radius=1.0):
    x = radius * math.cos(latitude) * math.cos(longitude)
    y = radius * math.cos(latitude) * math.sin(longitude)
    z = radius * math.sin(latitude)
    return np.array([x, y, z])


def rotate_vector(vector, yaw, pitch):
    """
    Rotates a vector by yaw (left/right) and pitch (up/down).
    This assumes:
      - yaw is a rotation around the Z-axis (vertical)
      - pitch is a rotation around the local X-axis (side-to-side)
    """
    # Rotation matrix for yaw around Z
    R_yaw = np.array([
        [math.cos(yaw), -math.sin(yaw), 0],
        [math.sin(yaw),  math.cos(yaw), 0],
        [0, 0, 1]
    ])
    
    # Rotation matrix for pitch around X
    R_pitch = np.array([
        [1, 0, 0],
        [0, math.cos(pitch), -math.sin(pitch)],
        [0, math.sin(pitch),  math.cos(pitch)]
    ])
    
    # Combine rotations: pitch then yaw
    R = R_yaw @ R_pitch
    
    return R @ vector




import numpy as np

def euler_to_rotation_matrix(yaw_deg, pitch_deg, roll_deg):
    # Convert degrees to radians
    yaw = np.radians(yaw_deg)
    pitch = np.radians(pitch_deg)
    roll = np.radians(roll_deg)

    # Rotation around X-axis (roll)
    Rx = np.array([
        [1, 0, 0],
        [0, np.cos(roll), -np.sin(roll)],
        [0, np.sin(roll),  np.cos(roll)]
    ])

    # Rotation around Y-axis (pitch)
    Ry = np.array([
        [ np.cos(pitch), 0, np.sin(pitch)],
        [0, 1, 0],
        [-np.sin(pitch), 0, np.cos(pitch)]
    ])

    # Rotation around Z-axis (yaw)
    Rz = np.array([
        [np.cos(yaw), -np.sin(yaw), 0],
        [np.sin(yaw),  np.cos(yaw), 0],
        [0, 0, 1]
    ])

    # Combined rotation matrix: R = Rz * Ry * Rx (yaw → pitch → roll)
    R = Rz @ Ry @ Rx
    return R









def local_to_global_gaze(opentrack_yaw, opentrack_pitch, opentrack_roll, person_longitude, person_latitude):
    # Get position of the person on the unit sphere


    person_position = spherical_to_cartesian(person_longitude, person_latitude,radius = 1)

    # Base gaze direction is inward (i.e., toward the origin)
    base_gaze_dir = -person_position / np.linalg.norm(person_position)
    

    # Convert yaw/pitch to radians
    yaw_rad = math.radians(opentrack_yaw)
    pitch_rad = math.radians(opentrack_pitch)

    # Rotate the inward gaze direction in local coordinate frame
    rotated = euler_to_rotation_matrix(opentrack_yaw, opentrack_pitch, opentrack_roll)

    

    final_gaze = rotated @ base_gaze_dir
    return tuple(final_gaze)
    # Step 1: Person's position on unit sphere
    # person_position = spherical_to_cartesian(person_longitude, person_latitude, radius=1)

    # # Step 2: Gaze vector is inward
    # base_gaze = -person_position / np.linalg.norm(person_position)

    # # Step 3: Construct rotation using yaw, pitch, roll (in degrees)
    # r = R.from_euler('ZYX', [opentrack_yaw, opentrack_pitch, opentrack_roll], degrees=True)

    # # Step 4: Apply local rotation to the base gaze direction
    # rotated_gaze = r.apply(base_gaze)

    # # Step 5: Convert to spherical → equirectangular image (2D)

    # return tuple(rotated_gaze)

   
   


#final_gaze_dir is a tuple 
#image is the global image to draw on 
def draw_arrow_global(image, start_pt, final_gaze_dir, color):
    cv2.arrowedLine(image,start_pt, final_gaze_dir, color, thickness = 2)

def cartesian_to_spherical(x, y, z):
    radius = math.sqrt(x**2 + y**2 + z**2)
    longitude = math.atan2(z,x)  # range: [-π, π]
    latitude = math.asin(y / radius)  # range: [-π/2, π/2]
    return longitude, latitude

def spherical_to_2d(longitude, latitude, frame_width, frame_height):
    # Clamp longitude to [-π, π]
    longitude = max(-math.pi, min(math.pi, longitude))
    
    # Clamp latitude to [-π/2, π/2]
    latitude = max(-math.pi/2, min(math.pi/2, latitude))
    
    # Convert to normalized coordinates [0, 1]
    u = (longitude + math.pi) / (2 * math.pi)
    v = (math.pi / 2 - latitude) / math.pi
    
    # Convert to pixel coordinates with bounds checking
    x = max(0, min(frame_width - 1, round(u * frame_width)))
    y = max(0, min(frame_height - 1, round(v * frame_height)))
    
    return x, y

File: src/c_impl/spherical_norm.cpp
#include <cmath>
#include <opencv2/opencv.hpp>

// Convert bbox center (pixel) to spherical coordinates (longitude, latitude)
std::pair<float, float> bbox_to_spherical_coord(float bbox_center_x, float bbox_center_y, int frame_width, int frame_height) {
    float u = bbox_center_x / frame_width;
    float v = bbox_center_y / frame_height;

    float longitude = (u * 2.0f * CV_PI) - CV_PI;        // [-π, π]
    float latitude = (CV_PI / 2.0f) - (v * CV_PI);      // [π/2, -π/2]

    return {longitude, latitude};
}

// Convert spherical to Cartesian coordinates (radius assumed)
cv::Vec3f spherical_to_cartesian(float longitude, float latitude, float radius) {
    float x = radius * cos(latitude) * cos(longitude);
    float y = radius * cos(latitude) * sin(longitude);
    float z = radius * sin(latitude);
    return cv::Vec3f(x, y, z);
}

// Rotate vector by yaw (around Z) and pitch (around X)
cv::Vec3f rotate_vector(const cv::Vec3f& vector, float yaw, float pitch) {
    // Rotation matrix for yaw (Z axis)
    cv::Matx33f R_yaw(
        cos(yaw), -sin(yaw), 0,
        sin(yaw),  cos(yaw), 0,
        0,        0,         1
    );

    // Rotation matrix for pitch (X axis)
    cv::Matx33f R_pitch(
        1, 0,           0,
        0, cos(pitch), -sin(pitch),
        0, sin(pitch),  cos(pitch)
    );

    // Combine rotations: pitch then yaw (R_yaw * R_pitch)
    cv::Matx33f R = R_yaw * R_pitch;

    cv::Vec3f rotated = R * vector;
    return rotated;
}

// Convert local gaze (yaw, pitch in degrees) and person position (longitude, latitude) 
// to global gaze direction vector
cv::Vec3f local_to_global_gaze(float opentrack_yaw, float opentrack_pitch, float person_longitude, float person_latitude) {
    // Person position on unit sphere
    cv::Vec3f person_pos = spherical_to_cartesian(person_longitude, person_latitude, 1.0f);

    // Base gaze direction is inward (towards origin)
    cv::Vec3f base_gaze_dir = -person_pos / cv::norm(person_pos);

    // Convert degrees to radians
    float yaw_rad = opentrack_yaw * CV_PI / 180.0f;
    float pitch_rad = opentrack_pitch * CV_PI / 180.0f;

    // Rotate gaze vector by yaw then pitch
    return rotate_vector(base_gaze_dir, yaw_rad, pitch_rad);
}

// Draw arrow from start_pt to end_pt (end_pt is a 2D point here)
void draw_arrow_global(cv::Mat& image, const cv::Point& start_pt, const cv::Point& end_pt, const cv::Scalar& color) {
    cv::arrowedLine(image, start_pt, end_pt, color, 2);
}

// Cartesian to spherical (longitude, latitude)
std::pair<float, float> cartesian_to_spherical(float x, float y, float z) {
    float radius = std::sqrt(x*x + y*y + z*z);
    float longitude = std::atan2(y, x);    // [-π, π]
    float latitude = std::asin(z / radius); // [-π/2, π/2]
    return {longitude, latitude};
}

// Spherical (longitude, latitude) to 2D pixel coordinates
cv::Point spherical_to_2d(float longitude, float latitude, int frame_width, int frame_height) {
    // Clamp
    if (longitude < -CV_PI) longitude = -CV_PI;
    else if (longitude > CV_PI) longitude = CV_PI;

    if (latitude < -CV_PI/2) latitude = -CV_PI/2;
    else if (latitude > CV_PI/2) latitude = CV_PI/2;

    float u = (longitude + CV_PI) / (2 * CV_PI);
    float v = (CV_PI / 2 - latitude) / CV_PI;

    int x = std::round(u * frame_width);
    int y = std::round(v * frame_height);

    // Clamp to frame bounds
    x = std::max(0, std::min(frame_width - 1, x));
    y = std::max(0, std::min(frame_height - 1, y));

    return cv::Point(x, y);
}
